src/java/com/puppetlabs/ssl_utils/SSLUtils.java

```diff
@@ -438,6 +445,60 @@ public static X509CRL revokeMultiple(X509CRL crl,
         return buildCRL(crl, issuerPrivateKey, issuerPublicKey, builder);
     }
 
+    /**
+     * Given a certificate chain and CRLs, validate CRLS and all certs in the chain.
+     *
+     * @param certs The certificate chain to validate
+     * @param crls The CRL chain to validate
+     * @throws CertificateException
+     * @throws CertPathValidatorException
+     * @throws IOException
+     * @throws InvalidAlgorithmParameterException
+     * @throws KeyStoreException
+     * @throws NoSuchAlgorithmException
+     * @throws NoSuchProviderException
+     */
+    public static void validateCertChain(List<X509Certificate> certs,
+                                       List<X509CRL> crls)
+        throws CertificateException, CertPathValidatorException, IOException,
+               InvalidAlgorithmParameterException, KeyStoreException,
+               NoSuchAlgorithmException, NoSuchProviderException {
+        // there are bouncycastle versions of CertificateFactory and CertPath
+        // that I haven't looked into - not sure when using bouncycastle vs. java
+        // makes sense
+        final CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+        final CertPath certPath = certFactory.generateCertPath(certs);
+        final CertPathValidator validator;
+        if (isFIPS()) {
+            validator = CertPathValidator.getInstance(PKIX_KEYMANAGER_ALGO, BOUNCYCASTLE_JSSE_PROVIDER);
```

Comment:
this is wrong, need to figure out the appropriate algorithm/provider combo for fips.
```
java.security.NoSuchAlgorithmException: no such algorithm: PKIX for provider BCJSSE
```

Reply:
This is weird, still looking into this, figured I should actually mention that I was doing that though ( ;

Reply:
okay, thanks! I was going to try using "BCFIPS" as the provider, since that was the only thing I found in my googling that seemed like it could maybe work haha. but I need Travis to test that for me, so I'll push something soon.

Reply:
Yeah, I see the BCFIPS being used in some examples. This just makes me confused because we should be configuring the BCJSSE to work in "fips" mode, ie, actually be backed by the BCFIPS provider. I don't know why BCJSSE wouldn't eventually defer to the BCFIPS provider here, or why if it doesn't do it here it would elsewhere... I have a feeling that it's an arcane java configuration issue or BC implementation quirk that would take forever to confirm.

Reply:
I don't really know anything about BCJSSE or providers in general - are you saying something is off somewhere that we need to fix?

Reply:
My understanding is that you should be able to say that you want the BCJSSE provider and then the java.security file we use should have it configured to actually be BCFIPS provider (or regular BC provider, depending on the profile you run the tests with).

I don't know why that doesn't work, it's probably just my understanding of it that's broken and it's probably a long road to fix that...

Reply:
ah, interesting. so would that behavior, in theory, come from the BCJSSE provider? like, it includes magic to detect the right underlying provider?

I'm noticing that the existing uses of the BCJSSE provider are only in FIPS mode. If I'm understanding what you're saying correctly, I would expect us to use that everywhere, rather than just for FIPS? I sorta just assumed that BCJSSE was a separate, unrelated provider. BCJSSE looks sorta specific to SSL stuff? https://www.bouncycastle.org/docs/tlsdocs1.5on/org/bouncycastle/jsse/package-summary.html

Reply:
I think section 2.1-3.1.2 of this explains it better than I can: https://downloads.bouncycastle.org/fips-java/BC-FJA-(D)TLSUserGuide-1.0.9.pdf

Reply:
I'll take a look on Monday 😜

Reply:
> The BC JSSE provider is built on top of the TLS API and provides functionality to support the use of the JSSE API defined in javax.net.ssl

Since the things I'm adding aren't in ^that java class, I think it makes sense that the BCJSSE provider doesn't work for this.